Design and interaction guidelines (#2398)
sourcegraph/about: handbook/product/design/design-and-interaction-guidelines.md

# Design and interaction guidelines

Sourcegraph’s search experience is the application’s primary interaction. As such, the design is tailored to the search use case by prioritizing code over other elements.

With the addition of campaigns, repository groups, and code insights, the product is quickly evolving. As such, its design must evolve to meet the experience requirements of these new features. 

This document will describe the current design and interaction models and describe the principles we will utilize to inform our design and interaction choices moving forward.

## Two design styles
Sourcegraph utilizes two design and interaction styles which are tailored to the content they display:

Search - An IDE like, full screen, high density, low contrast ui (for elements other than code) which centers on the display of code
Web application - a more traditional web application style with larger margins and padding, increased contrast and standard web page design experience requirements

## Code pages design

> In Sourcegraph’s search and file experiences, code is the most important element. The design of these pages reflects that priority. Contrast is balanced so that code stands out above all else. Panels clearly segment content and tools from the code. Pages are full width, and margins and padding are slim to increase the density of information which helps us display as much code on the screen as possible. 

Author: ```suggestion
In Sourcegraph’s search, file, diff and other views, code is the most important element. To collectively refer to these pages we use the phrase 'code pages'. The design of these pages should focus on code as the most important element. Contrast is balanced so that code stands out above all else. Panels clearly segment content and tools from the code. Pages are full width, and margins and padding are slim to increase the density of information which helps us display as much code on the screen as possible. 
```

The interaction model of the search experience is also unique. Clicking text on the page will almost always result in running a query or viewing a file. In a normal web application, we would clearly highlight these interactive elements so that the user understands where the actions are. If we were to do so on search pages, these elements would quickly outweigh and overwhelm the code. 

### Examples:

* [Search results](https://sourcegraph.com/search?q=repo:%5Egithub%5C.com/sourcegraph/sourcegraph%24+file:.*/go+auth&patternType=literal)
* [File views](https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/package.json)

<img src='./margins-example.png' />

_Small margins and reduced padding help to minimize scanning and scrolling. This allows developers to see code faster… and fit more of it on the screen._

> ### Icons and labels

Reviewer: I would consider changing the order here:

1. Code page design
2. Web app page design
3. Icons and labels - both for code page and web page -> it will be easier to compare those approached without jumping between the sections

@rrhyne @quinnkeast wdyt?

Author: Fine with me, would you mind making that change?

Author: I went to address that and realized I didn't understand what you meant. The icons and labels are sub-sections of each main topic (code pages and web pages) as they are treated in different ways for each. 

So in that sense, I don't think re-ordering makes sense. If Quinn does, we can address this in a separate pull request.

Reviewer: @quinnkeast wdyt?

<img src='./icons-example.png' />

_Extensions and file menus compared with the VS Code’s git and windowing functions._

## Web application page design

We increasingly need to render standard webpages that present larger bodies of text to the user, accept user input, or allow management of elements. These pages require page widths and margins that promote text readability, a clear distinction between links and body copy, and padding and margins that allow elements to state their importance and guide users through the content. 

### Example:

[Repository groups](https://sourcegraph.com/refactor-python2-to-3)
[Settings](https://sourcegraph.com/users/rrhyne/settings)
[Extensions](https://sourcegraph.com/extensions)
[Campaigns](https://k8s.sgdev.org/campaigns?visible=10)

<img src='./signup-example.png' />

_Clearly recognizable links help orient users to common signup actions._

<img src='./large-margin-example.png' />

_The large margins around the Kubernetes logo convey to the user: “This page is focused on Kubernetes”._

### Web application icons

> In the web application sections, icons should always have an accompanying label. See this article on [Icon Usability](https://www.nngroup.com/articles/icon-usability/) from the Nielson Norman group for more information. Note that this stance is contradictory to the requirements for icons in search application pages.

Reviewer: @rrhyne I wouldn't say it's ALWAYS necessary to have labels - for ex. 'info' icons with the tooltip on hover do not need labels. How about we change the wording to 'In the web application sections, icons should **generally** have an accompanying label.'?

Author: Good point.

Author: This should be resolved now, thanks for the suggestion.

> In the web application sections, icons should always have an accompanying label. See this article on [Icon Usability](https://www.nngroup.com/articles/icon-usability/) from the Nielson Norman group for more information. Note that this stance is contradictory to the requirements for icons in search application pages.

Author: ```suggestion
In the web application sections, icons should generally be accompanied by a label. See this article on [Icon Usability](https://www.nngroup.com/articles/icon-usability/) from the Nielson Norman group for more information. Note that this stance is contradictory to the requirements for icons in search application pages.
```
